### PR TITLE
Adding phase 1 API from other projects

### PR DESCRIPTION
**Pull request by embhorn**
* sk_free
* sk_pop_free
* X509_cmp
* X509_get_ext_count
* X509_STORE_CTX_get1_chain

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -15899,6 +15972,55 @@ void wolfSSL_sk_X509_free(WOLF_STACK_OF(WOLFSSL_X509_NAME)* sk) {
...
+   WOLFSSL_ENTER("wolfSSL_sk_ACCESS_DESCRIPTION_pop_free");
+
+   (void)f;
+   if (sk == NULL) {
+       return;
+   }
+
+   /* parse through stack freeing each node */
+   node = sk->next;
+   while (sk->num > 1) {
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: additional NULL check here on node

File: src/ssl.c [hunk 2]
```diff
@@ -16346,6 +16491,43 @@ void wolfSSL_sk_ASN1_OBJECT_free(WOLF_STACK_OF(WOLFSSL_ASN1_OBJECT)* sk)
...
+    WOLFSSL_ENTER("wolfSSL_sk_ASN1_OBJECT_pop_free");
+    (void)f;
+
+    if (sk == NULL) {
+        return;
+    }
+
+    /* parse through stack freeing each node */
+    node = sk->next;
+    while (sk->num > 1) {
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/ssl.c]: Check

File: src/ssl.c [hunk 3]
```diff
@@ -19221,6 +19500,89 @@ WOLFSSL_STACK* wolfSSL_X509_STORE_CTX_get_chain(WOLFSSL_X509_STORE_CTX* ctx)
...
+        }
+        XMEMSET(node, 0, sizeof(WOLFSSL_STACK));
+
+        /* copy sk node to new node, data by reference */
+        node->data.x509 = sIdx->data.x509;
+        node->num = sIdx->num;
+
+        /* insert node into list, progress idx */
+        if (i == 0) {
+            dup = node;
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/ssl.c]: dup is the head pointer. In the first iteration of the loop, dup->next should be NULL. dIdx saves node (on first iteration, set to dup). On second iteration, dIdx->next gets set to new node.

File: src/ssl.c [hunk 4]
```diff
@@ -19221,6 +19500,89 @@ WOLFSSL_STACK* wolfSSL_X509_STORE_CTX_get_chain(WOLFSSL_X509_STORE_CTX* ctx)
...
+        } else {
+            dIdx->next = node;
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/ssl.c]: It's used to store the end of the list, so the end->next can be set to the new node.

File: src/ssl.c (hunk 1 above)
Comment (embhorn) [reply src/ssl.c]: Check

File: src/ssl.c [hunk 5]
```diff
@@ -16346,6 +16491,43 @@ void wolfSSL_sk_ASN1_OBJECT_free(WOLF_STACK_OF(WOLFSSL_ASN1_OBJECT)* sk)
...
+        return;
+    }
+
+    /* parse through stack freeing each node */
+    node = sk->next;
+    while (sk->num > 1) {
+        WOLFSSL_STACK* tmp = node;
+        node = node->next;
+
+        f(tmp->data.obj);
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/ssl.c]: Check

File: src/ssl.c [hunk 6]
```diff
@@ -16346,6 +16491,43 @@ void wolfSSL_sk_ASN1_OBJECT_free(WOLF_STACK_OF(WOLFSSL_ASN1_OBJECT)* sk)
...
+        XFREE(tmp, NULL, DYNAMIC_TYPE_ASN1);
+        sk->num -= 1;
+    }
+
+    /* free head of stack */
+    if (sk->num == 1) {
+        f(sk->data.obj);
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/ssl.c]: Check

File: src/ssl.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (JacobBarthelmeh) [reply src/ssl.c]: You're right on this. dIdx get set as the tail and node->next gets set on second loop through.

File: src/ssl.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (JacobBarthelmeh) [reply src/ssl.c]: After writing a small test case looks like should actually fail in the case of NULL here

**Pull request merged**